File: src/main/java/com/atsuishio/superbwarfare/capability/ModCapabilities.java
package com.atsuishio.superbwarfare.capability;

import com.atsuishio.superbwarfare.Mod;
import com.atsuishio.superbwarfare.capability.player.PlayerVariable;
import dev.onyxstudios.cca.api.v3.component.ComponentKey;
import dev.onyxstudios.cca.api.v3.component.ComponentRegistry;
import dev.onyxstudios.cca.api.v3.entity.EntityComponentFactoryRegistry;
import dev.onyxstudios.cca.api.v3.entity.EntityComponentInitializer;
import dev.onyxstudios.cca.api.v3.entity.RespawnCopyStrategy;

public final class ModCapabilities implements EntityComponentInitializer {
    public static final ComponentKey<LaserCapability.ILaserCapability> LASER_CAPABILITY = ComponentRegistry.getOrCreate(Mod.loc("laser_capability"), LaserCapability.ILaserCapability.class);
    public static final ComponentKey<PlayerVariable> PLAYER_VARIABLE = ComponentRegistry.getOrCreate(Mod.loc("player_variables"), PlayerVariable.class);

    @Override
    public void registerEntityComponentFactories(EntityComponentFactoryRegistry registry) {
        registry.registerForPlayers(PLAYER_VARIABLE, player -> new PlayerVariable(), RespawnCopyStrategy.ALWAYS_COPY);
        registry.registerForPlayers(LASER_CAPABILITY, player -> new LaserCapability.LaserCapabilityImpl(), RespawnCopyStrategy.LOSSLESS_ONLY);
    }
}

File: src/main/kotlin/com/atsuishio/superbwarfare/capability/ModCapabilities.kt
package com.atsuishio.superbwarfare.capability

import com.atsuishio.superbwarfare.capability.entity.InfinityAmmoCapability
import com.atsuishio.superbwarfare.capability.living.PhosphorusFireCapability
import com.atsuishio.superbwarfare.capability.player.PlayerVariable
import dev.onyxstudios.cca.api.v3.component.ComponentKey
import dev.onyxstudios.cca.api.v3.component.ComponentRegistry
import dev.onyxstudios.cca.api.v3.entity.EntityComponentFactoryRegistry
import dev.onyxstudios.cca.api.v3.entity.EntityComponentInitializer
import dev.onyxstudios.cca.api.v3.entity.RespawnCopyStrategy
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.LivingEntity

class ModCapabilities : EntityComponentInitializer {

    override fun registerEntityComponentFactories(registry: EntityComponentFactoryRegistry) {
        registry.registerForPlayers(
            PLAYER_VARIABLE,
            { PlayerVariable() },
            RespawnCopyStrategy.ALWAYS_COPY
        )
        registry.registerFor(LivingEntity::class.java, PHOSPHORUS_FIRE) { PhosphorusFireCapability() }
        registry.registerFor(Entity::class.java, INFINITY_AMMO) { InfinityAmmoCapability() }
    }

    companion object {
        @JvmField
        val PLAYER_VARIABLE: ComponentKey<PlayerVariable> = ComponentRegistry.getOrCreate(
            PlayerVariable.ID,
            PlayerVariable::class.java
        )

        @JvmField
        val PHOSPHORUS_FIRE: ComponentKey<PhosphorusFireCapability> = ComponentRegistry.getOrCreate(
            PhosphorusFireCapability.ID,
            PhosphorusFireCapability::class.java
        )

        @JvmField
        val INFINITY_AMMO: ComponentKey<InfinityAmmoCapability> = ComponentRegistry.getOrCreate(
            InfinityAmmoCapability.ID,
            InfinityAmmoCapability::class.java
        )
    }
}
